### ISSUE TITLE
[tx] performance lowest hanging fruits

### ISSUE THREAD

**Issue by Guido1Alessandro1Trevisan** [NovaSky-AI/SkyRL#592]
I trained Qwen3-8B with the rl_loop on an H200, and I noticed that it can’t handle large batch sizes — after 4 microbatch accumulations with group size = 8, it crashes when using batch_size = 8. It’s also relatively slow. #591 

I was thinking about stuff like gradient checkpointing and ep. Any ideas on quick / medium fixes perfomance-wise?

**Comment by pcmoritz** [comment]
One of them is definitely https://github.com/NovaSky-AI/SkyRL/pull/590, due to that the lm_head was not properly sharded (maybe others too), with the fix you can run larger batch sizes. I'll merge it soon :)

**Comment by Guido1Alessandro1Trevisan** [comment]
Awesome! Looks like it’s merged. If there are any more you think would be worth fixing I’m ready to jump on them. I'll think of some myself.

**Comment by pcmoritz** [comment]
Sounds great, yes that will be great! If you have time, it would be great to try the jax profiler https://docs.jax.dev/en/latest/profiling.html and see where the bottlenecks are at the moment. If you share them here I'm happy to help try to figure out how to fix them :)

Right now my focus is on getting the 0.1 release out, that means running some more RL experiments and having more loss curves and then writing the blog post. I can help with the profiling / performance fixing after :)

**Comment by pcmoritz** [comment]
It might also be a good idea to try a CPU profiler like py-spy, I saw there are sometimes phases where the GPU is not utilized at all, I have a feeling there is some low hanging fruit in optimizing some CPU stuff too :)

**Comment by pcmoritz** [comment]
It seems currently a decent amount of time is being spent in _accumulate_grads (this is for a 4B model), I wonder if we can speed that up by jitting it (or a subset of it) :)

**Comment by pcmoritz** [comment]
I made a fix for _accumulate_grads in https://github.com/NovaSky-AI/SkyRL/pull/614 btw :)

**Comment by pcmoritz** [comment]
There is more to be optimized for accumulate grads, I created an issue here: https://github.com/NovaSky-AI/SkyRL/issues/647
